### ISSUE TITLE
connection reset by peer - channel robustness in python

### ISSUE THREAD

**Issue by tilsche** [metricq/metricq#38]
Hapens to grafana sometimes:

```
Aug 20 10:05:04 igel metricq-grafana[3302]: 2019-08-20 10:05:04,594 [INFO    ] [aiohttp.access      ] 127.0.0.1 [20/Aug/2019:08:04:20 +0000] "POST /query HTTP/1.1" 404 191 "https://grafana.z
Aug 20 10:05:04 igel metricq-grafana[3302]: 2019-08-20 10:05:04,594 [INFO    ] [metricq.agent       ] closing connection to [Errno 104] Connection reset by peer
Aug 20 10:05:04 igel metricq-grafana[3302]: 2019-08-20 10:05:04,603 [ERROR   ] [aiohttp.server      ] Error handling request
Aug 20 10:05:04 igel metricq-grafana[3302]: Traceback (most recent call last):
Aug 20 10:05:04 igel metricq-grafana[3302]:   File "/home/service/envs/grafana/lib/python3.7/site-packages/aiohttp/web_protocol.py", line 418, in start
Aug 20 10:05:04 igel metricq-grafana[3302]:     resp = await task
Aug 20 10:05:04 igel metricq-grafana[3302]:   File "/home/service/envs/grafana/lib/python3.7/site-packages/aiohttp/web_app.py", line 458, in _handle
Aug 20 10:05:04 igel metricq-grafana[3302]:     resp = await handler(request)
Aug 20 10:05:04 igel metricq-grafana[3302]:   File "/home/service/metricq-grafana/metricq_grafana/views.py", line 19, in query
Aug 20 10:05:04 igel metricq-grafana[3302]:     resp = await get_history_data(request.app, req_json)
Aug 20 10:05:04 igel metricq-grafana[3302]:   File "/home/service/metricq-grafana/metricq_grafana/amqp.py", line 36, in get_history_data
Aug 20 10:05:04 igel metricq-grafana[3302]:     for target in targets
Aug 20 10:05:04 igel metricq-grafana[3302]:   File "/home/service/metricq-grafana/metricq_grafana/target.py", line 45, in get_response
Aug 20 10:05:04 igel metricq-grafana[3302]:     self._get_data(app, start_time, end_time, interval), self._get_metadata(app)
Aug 20 10:05:04 igel metricq-grafana[3302]:   File "/home/service/metricq-grafana/metricq_grafana/target.py", line 73, in _get_data
Aug 20 10:05:04 igel metricq-grafana[3302]:     self.metric, start_time, end_time, interval, timeout=30
Aug 20 10:05:04 igel metricq-grafana[3302]:   File "/home/service/metricq/python/metricq/history_client.py", line 264, in history_data_request
Aug 20 10:05:04 igel metricq-grafana[3302]:     await self.history_exchange.publish(msg, metric)
Aug 20 10:05:04 igel metricq-grafana[3302]:   File "/home/service/envs/grafana/lib/python3.7/site-packages/aio_pika/exchange.py", line 202, in publish
Aug 20 10:05:04 igel metricq-grafana[3302]:     loop=self.loop, timeout=timeout
Aug 20 10:05:04 igel metricq-grafana[3302]:   File "/home/service/envs/grafana/lib/python3.7/asyncio/tasks.py", line 388, in wait_for
Aug 20 10:05:04 igel metricq-grafana[3302]:     return await fut
Aug 20 10:05:04 igel metricq-grafana[3302]:   File "/home/service/envs/grafana/lib/python3.7/site-packages/aiormq/channel.py", line 398, in basic_publish
Aug 20 10:05:04 igel metricq-grafana[3302]:     async with self.lock:
Aug 20 10:05:04 igel metricq-grafana[3302]:   File "/home/service/envs/grafana/lib/python3.7/site-packages/aiormq/channel.py", line 73, in lock
Aug 20 10:05:04 igel metricq-grafana[3302]:     raise RuntimeError('%r closed' % self)
Aug 20 10:05:04 igel metricq-grafana[3302]: RuntimeError: <Channel: "1"> closed

```

The error repeats often, and there's an intermediate:

```
Aug 20 10:05:09 igel metricq-grafana[3302]: 2019-08-20 10:05:09,725 [ERROR   ] [aio_pika.robust_connection] Open channel failure
Aug 20 10:05:09 igel metricq-grafana[3302]: Traceback (most recent call last):
Aug 20 10:05:09 igel metricq-grafana[3302]:   File "/home/service/envs/grafana/lib/python3.7/site-packages/aiormq/base.py", line 25, in __inner
Aug 20 10:05:09 igel metricq-grafana[3302]:     return await self.task
Aug 20 10:05:09 igel metricq-grafana[3302]: concurrent.futures._base.CancelledError
Aug 20 10:05:09 igel metricq-grafana[3302]: The above exception was the direct cause of the following exception:
Aug 20 10:05:09 igel metricq-grafana[3302]: Traceback (most recent call last):
Aug 20 10:05:09 igel metricq-grafana[3302]:   File "/home/service/envs/grafana/lib/python3.7/site-packages/aio_pika/robust_connection.py", line 107, in _on_reconnect
Aug 20 10:05:09 igel metricq-grafana[3302]:     await channel.on_reconnect(self, number)
Aug 20 10:05:09 igel metricq-grafana[3302]:   File "/home/service/envs/grafana/lib/python3.7/site-packages/aio_pika/robust_channel.py", line 62, in on_reconnect
Aug 20 10:05:09 igel metricq-grafana[3302]:     await queue.on_reconnect(self)
Aug 20 10:05:09 igel metricq-grafana[3302]:   File "/home/service/envs/grafana/lib/python3.7/site-packages/aio_pika/robust_queue.py", line 50, in on_reconnect
Aug 20 10:05:09 igel metricq-grafana[3302]:     await self.declare()
Aug 20 10:05:09 igel metricq-grafana[3302]:   File "/home/service/envs/grafana/lib/python3.7/site-packages/aio_pika/queue.py", line 86, in declare
Aug 20 10:05:09 igel metricq-grafana[3302]:     ), timeout=timeout, loop=self.loop
Aug 20 10:05:09 igel metricq-grafana[3302]:   File "/home/service/envs/grafana/lib/python3.7/asyncio/tasks.py", line 388, in wait_for
Aug 20 10:05:09 igel metricq-grafana[3302]:     return await fut
Aug 20 10:05:09 igel metricq-grafana[3302]:   File "/home/service/envs/grafana/lib/python3.7/site-packages/aiormq/channel.py", line 531, in queue_declare
Aug 20 10:05:09 igel metricq-grafana[3302]:     arguments=arguments,
Aug 20 10:05:09 igel metricq-grafana[3302]:   File "/home/service/envs/grafana/lib/python3.7/site-packages/aiormq/base.py", line 171, in wrap
Aug 20 10:05:09 igel metricq-grafana[3302]:     return await self.create_task(func(self, *args, **kwargs))
Aug 20 10:05:09 igel metricq-grafana[3302]:   File "/home/service/envs/grafana/lib/python3.7/site-packages/aiormq/base.py", line 27, in __inner
Aug 20 10:05:09 igel metricq-grafana[3302]:     raise self.exception from e
Aug 20 10:05:09 igel metricq-grafana[3302]:   File "/home/service/envs/grafana/lib/python3.7/asyncio/tasks.py", line 603, in _wrap_awaitable
Aug 20 10:05:09 igel metricq-grafana[3302]:     return (yield from awaitable.__await__())
Aug 20 10:05:09 igel metricq-grafana[3302]:   File "/home/service/envs/grafana/lib/python3.7/site-packages/aiormq/base.py", line 27, in __inner
Aug 20 10:05:09 igel metricq-grafana[3302]:     raise self.exception from e
Aug 20 10:05:09 igel metricq-grafana[3302]: aiormq.exceptions.ChannelNotFoundEntity: NOT_FOUND - no queue 'history-grafana-hrsp' in vhost '/'
```

This seems to be somewhat different than #23 / #11 - but the NOT FOUND may be similar.. not exactly sure what's happening here. Maybe it is trying to reconnect, but failed eventually because the manager-created queue doesn't exist any longer. Maybe we just need to use the C++ approach of stopping on error and letting systemd restart the daemon (except for sources, who don't have managed queues).

**Comment by tilsche** [comment]
On RabbitMQ side:

    2019-08-21 10:44:04.417 [error] <0.4410.125> closing AMQP connection <0.4410.125> (xx.xx.xx.xx:44646 -> xx.xx.xx.xx:5671):
{writer,send_failed,timeout}

**Comment by phijor** [comment]
> The error repeats often, and there's an intermediate:
> 
> ```
> Aug 20 10:05:09 igel metricq-grafana[3302]: 2019-08-20 10:05:09,725 [ERROR   ] [aio_pika.robust_connection] Open channel failure
> Aug 20 10:05:09 igel metricq-grafana[3302]: Traceback (most recent call last):
> Aug 20 10:05:09 igel metricq-grafana[3302]:   File "/home/service/envs/grafana/lib/python3.7/site-packages/aiormq/base.py", line 25, in __inner
> Aug 20 10:05:09 igel metricq-grafana[3302]:     return await self.task
> Aug 20 10:05:09 igel metricq-grafana[3302]: concurrent.futures._base.CancelledError
> Aug 20 10:05:09 igel metricq-grafana[3302]: The above exception was the direct cause of the following exception:
> Aug 20 10:05:09 igel metricq-grafana[3302]: Traceback (most recent call last):
> Aug 20 10:05:09 igel metricq-grafana[3302]:   File "/home/service/envs/grafana/lib/python3.7/site-packages/aio_pika/robust_connection.py", line 107, in _on_reconnect
> Aug 20 10:05:09 igel metricq-grafana[3302]:     await channel.on_reconnect(self, number)
> Aug 20 10:05:09 igel metricq-grafana[3302]:   File "/home/service/envs/grafana/lib/python3.7/site-packages/aio_pika/robust_channel.py", line 62, in on_reconnect
> Aug 20 10:05:09 igel metricq-grafana[3302]:     await queue.on_reconnect(self)
> Aug 20 10:05:09 igel metricq-grafana[3302]:   File "/home/service/envs/grafana/lib/python3.7/site-packages/aio_pika/robust_queue.py", line 50, in on_reconnect
> Aug 20 10:05:09 igel metricq-grafana[3302]:     await self.declare()
> Aug 20 10:05:09 igel metricq-grafana[3302]:   File "/home/service/envs/grafana/lib/python3.7/site-packages/aio_pika/queue.py", line 86, in declare
> Aug 20 10:05:09 igel metricq-grafana[3302]:     ), timeout=timeout, loop=self.loop
> Aug 20 10:05:09 igel metricq-grafana[3302]:   File "/home/service/envs/grafana/lib/python3.7/asyncio/tasks.py", line 388, in wait_for
> Aug 20 10:05:09 igel metricq-grafana[3302]:     return await fut
> Aug 20 10:05:09 igel metricq-grafana[3302]:   File "/home/service/envs/grafana/lib/python3.7/site-packages/aiormq/channel.py", line 531, in queue_declare
> Aug 20 10:05:09 igel metricq-grafana[3302]:     arguments=arguments,
> Aug 20 10:05:09 igel metricq-grafana[3302]:   File "/home/service/envs/grafana/lib/python3.7/site-packages/aiormq/base.py", line 171, in wrap
> Aug 20 10:05:09 igel metricq-grafana[3302]:     return await self.create_task(func(self, *args, **kwargs))
> Aug 20 10:05:09 igel metricq-grafana[3302]:   File "/home/service/envs/grafana/lib/python3.7/site-packages/aiormq/base.py", line 27, in __inner
> Aug 20 10:05:09 igel metricq-grafana[3302]:     raise self.exception from e
> Aug 20 10:05:09 igel metricq-grafana[3302]:   File "/home/service/envs/grafana/lib/python3.7/asyncio/tasks.py", line 603, in _wrap_awaitable
> Aug 20 10:05:09 igel metricq-grafana[3302]:     return (yield from awaitable.__await__())
> Aug 20 10:05:09 igel metricq-grafana[3302]:   File "/home/service/envs/grafana/lib/python3.7/site-packages/aiormq/base.py", line 27, in __inner
> Aug 20 10:05:09 igel metricq-grafana[3302]:     raise self.exception from e
> Aug 20 10:05:09 igel metricq-grafana[3302]: aiormq.exceptions.ChannelNotFoundEntity: NOT_FOUND - no queue 'history-grafana-hrsp' in vhost '/'
> ```

I couldn't yet reproduce the first error, but this one is caused and logged by `aio_pika` itself:

https://github.com/mosquito/aio-pika/blob/728869b2b83501955026f623963fa01de6b93f6d/aio_pika/robust_connection.py#L125-L134

Down the stack it tries to re-declare the manager-assigned queue:

https://github.com/mosquito/aio-pika/blob/728869b2b83501955026f623963fa01de6b93f6d/aio_pika/robust_queue.py#L50

Seems like we can't really do anything about it, except for writing our own `RobustConnection`.

**Comment by tilsche** [comment]
It is clear that we can't do anything about transient manager-assigned queues

**Comment by bmario** [comment]
Shouldn't that be fixed with v1.1.0 now?

**Comment by phijor** [comment]
Closed in #46.
